File: backend/seed.py
"""
Seed script — run with:  python seed.py
Creates sample departments, users, employees, leave requests, and notifications.
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from datetime import date, datetime, timezone
from app.core.database import SessionLocal
from app.core.security import hash_password
from app.models.user import User
from app.models.department import Department
from app.models.employee import Employee
from app.models.leave_request import LeaveRequest
from app.models.notification import Notification
from app.models.attendance import Attendance
from app.models.recruitment import Job


def seed():
    db = SessionLocal()
    try:
        print("🌱  Seeding database …")

        # ── Guard: skip if already seeded ─────────────────────────────────────
        if db.query(User).count() > 0:
            print("⚠️   Database already seeded — skipping.")
            return

        # ── Departments ───────────────────────────────────────────────────────
        depts = {
            "Engineering":     Department(name="Engineering"),
            "Human Resources": Department(name="Human Resources"),
            "Finance":         Department(name="Finance"),
            "Marketing":       Department(name="Marketing"),
            "Operations":      Department(name="Operations"),
        }
        for d in depts.values():
            db.add(d)
        db.flush()
        print(f"   ✔ {len(depts)} departments")

        # ── Admin user ────────────────────────────────────────────────────────
        admin_user = User(
            name="HR Admin",
            email="admin@hrportal.com",
            password=hash_password("Admin@1234"),
            role="admin",
            is_active=True,
        )
        db.add(admin_user)
        db.flush()

        # ── Employees (with linked user accounts) ─────────────────────────────
        employees_data = [
            ("Alice Johnson",  "alice.johnson@company.com",  "alice",  "Senior Engineer",     95000, depts["Engineering"],     date(2021, 3, 15)),
            ("Bob Smith",      "bob.smith@company.com",      "bob",    "HR Specialist",        62000, depts["Human Resources"], date(2020, 7, 1)),
            ("Carol White",    "carol.white@company.com",    "carol",  "Financial Analyst",    75000, depts["Finance"],         date(2022, 1, 10)),
            ("David Lee",      "david.lee@company.com",      "david",  "Marketing Manager",    85000, depts["Marketing"],       date(2019, 11, 5)),
            ("Eva Martinez",   "eva.martinez@company.com",   "eva",    "DevOps Engineer",      90000, depts["Engineering"],     date(2023, 4, 20)),
            ("Frank Brown",    "frank.brown@company.com",    "frank",  "Operations Analyst",   68000, depts["Operations"],      date(2021, 9, 12)),
        ]

        employee_objects: list[Employee] = []
        for name, email, slug, title, salary, dept, start in employees_data:
            user = User(
                name=name,
                email=email,
                password=hash_password("Password@1234"),
                role="employee",
                is_active=True,
            )
            db.add(user)
            db.flush()

            emp = Employee(
                name=name,
                email=email,
                title=title,
                salary=salary,
                status="active",
                start_date=start,
                department_id=dept.id,
                user_id=user.id,
            )
            db.add(emp)
            db.flush()
            employee_objects.append(emp)

        print(f"   ✔ {len(employee_objects)} employees + user accounts")

        # ── Leave requests ────────────────────────────────────────────────────
        leave_data = [
            (employee_objects[0], "annual_leave",    date(2025, 8, 1),  date(2025, 8, 7),  "Family vacation",   "pending"),
            (employee_objects[1], "sick_leave",      date(2025, 7, 15), date(2025, 7, 17), "Doctor appointment","approved"),
            (employee_objects[2], "emergency_leave", date(2025, 7, 20), date(2025, 7, 21), "Family emergency",  "rejected"),
        ]
        for emp, ltype, start, end, reason, status in leave_data:
            lr = LeaveRequest(
                employee_id=emp.id,
                leave_type=ltype,
                start_date=start,
                end_date=end,
                reason=reason,
                status=status,
                approved_by_user_id=admin_user.id if status != "pending" else None,
                admin_comment="Approved" if status == "approved" else ("Not feasible at this time" if status == "rejected" else None),
            )
            db.add(lr)
        print(f"   ✔ {len(leave_data)} leave requests")

        # ── Notifications ─────────────────────────────────────────────────────
        notifs = [
            Notification(
                type="announcement",
                title="Welcome to the HR Portal",
                message="We are excited to launch our new HR management system. Please explore all features and reach out with any questions.",
                audience_role=None,
                is_published=True,
                published_at=datetime.now(timezone.utc),
                created_by_user_id=admin_user.id,
            ),
            Notification(
                type="holiday",
                title="Public Holiday — Independence Day",
                message="The office will be closed on July 4th. Enjoy the long weekend!",
                audience_role=None,
                is_published=True,
                published_at=datetime.now(timezone.utc),
                created_by_user_id=admin_user.id,
            ),
            Notification(
                type="policy_update",
                title="Updated Leave Policy",
                message="Our leave policy has been updated. Annual leave now includes 2 extra days for employees with 3+ years tenure. Full details in the HR portal.",
                audience_role="employee",
                is_published=True,
                published_at=datetime.now(timezone.utc),
                created_by_user_id=admin_user.id,
            ),
        ]
        for n in notifs:
            db.add(n)
        print(f"   ✔ {len(notifs)} notifications")

        # ── Open jobs ─────────────────────────────────────────────────────────
        jobs = [
            Job(
                title="Senior Full-Stack Developer",
                description="We are looking for an experienced full-stack developer to join our engineering team.",
                location="Remote",
                status="open",
                opened_at=datetime.now(timezone.utc),
                department_id=depts["Engineering"].id,
            ),
            Job(
                title="HR Business Partner",
                description="Join our HR team to help drive people strategies across the organization.",
                location="On-site",
                status="open",
                opened_at=datetime.now(timezone.utc),
                department_id=depts["Human Resources"].id,
            ),
        ]
        for j in jobs:
            db.add(j)
        print(f"   ✔ {len(jobs)} open jobs")

        db.commit()
        print("\n✅  Seed complete!")
        print("\n📋  Login credentials:")
        print("   Admin    → admin@hrportal.com    / Admin@1234")
        print("   Employee → alice.johnson@company.com / Password@1234")
        print("   (all 6 employees share the same password: Password@1234)\n")

    except Exception as e:
        db.rollback()
        print(f"\n❌  Seed failed: {e}")
        raise
    finally:
        db.close()


if __name__ == "__main__":
    seed()

File: backend/prisma/seed.ts
import { PrismaClient, Prisma } from '@prisma/client';
import bcrypt from 'bcrypt';

const prisma = new PrismaClient();

async function main() {
  console.log('🌱 Seeding database...\n');

  // ── Departments ──────────────────────────────────────────────────────────────

  const departments = await Promise.all([
    prisma.department.upsert({
      where: { name: 'Engineering' },
      update: {},
      create: { name: 'Engineering' },
    }),
    prisma.department.upsert({
      where: { name: 'Human Resources' },
      update: {},
      create: { name: 'Human Resources' },
    }),
    prisma.department.upsert({
      where: { name: 'Finance' },
      update: {},
      create: { name: 'Finance' },
    }),
    prisma.department.upsert({
      where: { name: 'Marketing' },
      update: {},
      create: { name: 'Marketing' },
    }),
    prisma.department.upsert({
      where: { name: 'Product' },
      update: {},
      create: { name: 'Product' },
    }),
  ]);

  const [engineering, hr, finance, marketing, product] = departments;
  console.log(`✅ Upserted ${departments.length} departments`);

  // ── Admin user ────────────────────────────────────────────────────────────────

  const adminPassword = await bcrypt.hash('Admin@1234', 12);
  const adminUser = await prisma.user.upsert({
    where: { email: 'admin@hrportal.com' },
    update: {},
    create: {
      name: 'HR Admin',
      email: 'admin@hrportal.com',
      password: adminPassword,
      role: 'admin',
    },
  });
  console.log(`✅ Admin user: admin@hrportal.com / Admin@1234`);

  // ── Employee users ────────────────────────────────────────────────────────────

  const employeePassword = await bcrypt.hash('Password@1234', 12);

  const employeeUsers = await Promise.all([
    prisma.user.upsert({
      where: { email: 'avery.morgan@company.com' },
      update: {},
      create: { name: 'Avery Morgan', email: 'avery.morgan@company.com', password: employeePassword, role: 'employee' },
    }),
    prisma.user.upsert({
      where: { email: 'jordan.blake@company.com' },
      update: {},
      create: { name: 'Jordan Blake', email: 'jordan.blake@company.com', password: employeePassword, role: 'employee' },
    }),
    prisma.user.upsert({
      where: { email: 'sam.rivera@company.com' },
      update: {},
      create: { name: 'Sam Rivera', email: 'sam.rivera@company.com', password: employeePassword, role: 'employee' },
    }),
    prisma.user.upsert({
      where: { email: 'taylor.kim@company.com' },
      update: {},
      create: { name: 'Taylor Kim', email: 'taylor.kim@company.com', password: employeePassword, role: 'employee' },
    }),
    prisma.user.upsert({
      where: { email: 'riley.chen@company.com' },
      update: {},
      create: { name: 'Riley Chen', email: 'riley.chen@company.com', password: employeePassword, role: 'employee' },
    }),
    prisma.user.upsert({
      where: { email: 'casey.patel@company.com' },
      update: {},
      create: { name: 'Casey Patel', email: 'casey.patel@company.com', password: employeePassword, role: 'employee' },
    }),
  ]);
  console.log(`✅ Created ${employeeUsers.length} employee users`);

  // ── Employee profiles ─────────────────────────────────────────────────────────

  const employeeProfiles = await Promise.all([
    prisma.employee.upsert({
      where: { email: 'avery.morgan@company.com' },
      update: {},
      create: {
        name: 'Avery Morgan',
        email: 'avery.morgan@company.com',
        title: 'Senior Software Engineer',
        salary: new Prisma.Decimal(95000),
        status: 'active',
        start_date: new Date('2022-03-15'),
        department_id: engineering.id,
        user_id: employeeUsers[0].id,
      },
    }),
    prisma.employee.upsert({
      where: { email: 'jordan.blake@company.com' },
      update: {},
      create: {
        name: 'Jordan Blake',
        email: 'jordan.blake@company.com',
        title: 'HR Manager',
        salary: new Prisma.Decimal(78000),
        status: 'active',
        start_date: new Date('2021-07-01'),
        department_id: hr.id,
        user_id: employeeUsers[1].id,
      },
    }),
    prisma.employee.upsert({
      where: { email: 'sam.rivera@company.com' },
      update: {},
      create: {
        name: 'Sam Rivera',
        email: 'sam.rivera@company.com',
        title: 'Financial Analyst',
        salary: new Prisma.Decimal(72000),
        status: 'active',
        start_date: new Date('2023-01-10'),
        department_id: finance.id,
        user_id: employeeUsers[2].id,
      },
    }),
    prisma.employee.upsert({
      where: { email: 'taylor.kim@company.com' },
      update: {},
      create: {
        name: 'Taylor Kim',
        email: 'taylor.kim@company.com',
        title: 'Marketing Specialist',
        salary: new Prisma.Decimal(68000),
        status: 'active',
        start_date: new Date('2022-11-01'),
        department_id: marketing.id,
        user_id: employeeUsers[3].id,
      },
    }),
    prisma.employee.upsert({
      where: { email: 'riley.chen@company.com' },
      update: {},
      create: {
        name: 'Riley Chen',
        email: 'riley.chen@company.com',
        title: 'Product Manager',
        salary: new Prisma.Decimal(105000),
        status: 'active',
        start_date: new Date('2020-05-20'),
        department_id: product.id,
        user_id: employeeUsers[4].id,
      },
    }),
    prisma.employee.upsert({
      where: { email: 'casey.patel@company.com' },
      update: {},
      create: {
        name: 'Casey Patel',
        email: 'casey.patel@company.com',
        title: 'Junior Developer',
        salary: new Prisma.Decimal(58000),
        status: 'active',
        start_date: new Date('2024-02-01'),
        department_id: engineering.id,
        user_id: employeeUsers[5].id,
      },
    }),
  ]);
  console.log(`✅ Created ${employeeProfiles.length} employee profiles`);

  // ── Leave requests ────────────────────────────────────────────────────────────

  await prisma.leaveRequest.createMany({
    skipDuplicates: true,
    data: [
      {
        employee_id: employeeProfiles[0].id,
        leave_type: 'annual_leave',
        start_date: new Date('2025-08-01'),
        end_date: new Date('2025-08-07'),
        reason: 'Family vacation',
        status: 'pending',
      },
      {
        employee_id: employeeProfiles[1].id,
        leave_type: 'sick_leave',
        start_date: new Date('2025-07-15'),
        end_date: new Date('2025-07-16'),
        reason: 'Medical appointment',
        status: 'approved',
        approved_by_user_id: adminUser.id,
        admin_comment: 'Approved. Get well soon.',
      },
      {
        employee_id: employeeProfiles[2].id,
        leave_type: 'emergency_leave',
        start_date: new Date('2025-07-20'),
        end_date: new Date('2025-07-22'),
        reason: 'Family emergency',
        status: 'pending',
      },
    ],
  });
  console.log('✅ Created sample leave requests');

  // ── Attendance records (today + last 7 days) ──────────────────────────────────

  const now = new Date();
  for (let i = 0; i < 5; i++) {
    const workDate = new Date(now);
    workDate.setDate(now.getDate() - i);
    workDate.setUTCHours(0, 0, 0, 0);

    // Skip weekends
    const dayOfWeek = workDate.getDay();
    if (dayOfWeek === 0 || dayOfWeek === 6) continue;

    for (const employee of employeeProfiles.slice(0, 4)) {
      const checkIn = new Date(workDate);
      checkIn.setHours(8 + Math.floor(Math.random() * 2), Math.floor(Math.random() * 30), 0, 0);

      const checkOut = new Date(workDate);
      checkOut.setHours(17 + Math.floor(Math.random() * 2), Math.floor(Math.random() * 30), 0, 0);

      await prisma.attendance.upsert({
        where: {
          employee_id_work_date: {
            employee_id: employee.id,
            work_date: workDate,
          },
        },
        update: {},
        create: {
          employee_id: employee.id,
          work_date: workDate,
          check_in_at: checkIn,
          check_out_at: i === 0 ? undefined : checkOut,
          status: checkIn.getHours() > 9 ? 'late' : 'present',
        },
      });
    }
  }
  console.log('✅ Created attendance records');

  // ── Notifications ─────────────────────────────────────────────────────────────

  await prisma.notification.createMany({
    skipDuplicates: true,
    data: [
      {
        type: 'announcement',
        title: 'Welcome to HR Portal!',
        message: 'We are excited to launch our new HR management system. Please explore all features.',
        is_published: true,
        published_at: new Date(),
        created_by_user_id: adminUser.id,
      },
      {
        type: 'holiday',
        title: 'Public Holiday Notice',
        message: 'The office will be closed on July 4th for Independence Day.',
        is_published: true,
        published_at: new Date(),
        created_by_user_id: adminUser.id,
      },
      {
        type: 'policy_update',
        title: 'Updated Leave Policy',
        message: 'Our leave policy has been updated. Annual leave is now 20 days per year.',
        is_published: true,
        published_at: new Date(),
        created_by_user_id: adminUser.id,
      },
    ],
  });
  console.log('✅ Created notifications');

  // ── Jobs ──────────────────────────────────────────────────────────────────────

  const jobs = await Promise.all([
    prisma.job.upsert({
      where: { id: 'seed-job-001-0000-0000-000000000001' },
      update: {},
      create: {
        id: 'seed-job-001-0000-0000-000000000001',
        title: 'Full Stack Developer',
        description: 'We are looking for a Full Stack Developer proficient in React and Node.js to join our engineering team.',
        location: 'Remote',
        department_id: engineering.id,
        status: 'open',
        opened_at: new Date(),
      },
    }),
    prisma.job.upsert({
      where: { id: 'seed-job-002-0000-0000-000000000002' },
      update: {},
      create: {
        id: 'seed-job-002-0000-0000-000000000002',
        title: 'Product Designer',
        description: 'Looking for a creative Product Designer to craft exceptional user experiences.',
        location: 'New York, NY',
        department_id: product.id,
        status: 'open',
        opened_at: new Date(),
      },
    }),
  ]);
  console.log('✅ Created job postings');

  // ── Candidates ────────────────────────────────────────────────────────────────

  await prisma.candidate.createMany({
    skipDuplicates: true,
    data: [
      {
        job_id: jobs[0].id,
        name: 'Alex Turner',
        email: 'alex.turner@email.com',
        phone: '+1-555-0101',
        status: 'screening',
        notes: 'Strong React skills. Schedule technical interview.',
      },
      {
        job_id: jobs[0].id,
        name: 'Morgan Lee',
        email: 'morgan.lee@email.com',
        phone: '+1-555-0102',
        status: 'applied',
      },
      {
        job_id: jobs[1].id,
        name: 'Jamie Wilson',
        email: 'jamie.wilson@email.com',
        phone: '+1-555-0103',
        status: 'interview_scheduled',
        interview_at: new Date('2025-08-10T14:00:00Z'),
        notes: 'Great portfolio. Final round interview.',
      },
    ],
  });
  console.log('✅ Created candidates\n');

  console.log('─────────────────────────────────────────');
  console.log('🎉 Seeding complete!\n');
  console.log('Login credentials:');
  console.log('  Admin   : admin@hrportal.com     / Admin@1234');
  console.log('  Employee: avery.morgan@company.com / Password@1234');
  console.log('─────────────────────────────────────────\n');
}

main()
  .catch((e) => {
    console.error('❌ Seed failed:', e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
